File: src/test/java/de/espend/idea/php/annotation/tests/navigation/AnnotationUsageLineMarkerProviderTest.java
package de.espend.idea.php.annotation.tests.navigation;

import com.jetbrains.php.lang.psi.PhpPsiElementFactory;
import de.espend.idea.php.annotation.tests.AnnotationLightCodeInsightFixtureTestCase;

/**
 * @author Daniel Espendiller <daniel@espendiller.net>
 */
public class AnnotationUsageLineMarkerProviderTest extends AnnotationLightCodeInsightFixtureTestCase {
    public void setUp() throws Exception {
        super.setUp();
        myFixture.copyFileToProject("AnnotationUsageLineMarkerProvider.php");
    }

    public String getTestDataPath() {
        return "src/test/java/de/espend/idea/php/annotation/tests/navigation/fixtures";
    }

    public void testThatLineMarkerIsProvidedForAnnotationClass() {
        assertLineMarker(PhpPsiElementFactory.createPsiFileFromText(getProject(), "<?php\n" +
            "namespace Doctrine\\ORM\\Mapping;\n" +
            "" +
            "{\n" +
            "   /**\n" +
            "   * @Annotation\n" +
            "   */\n" +
            "   class Embedded\n" +
            "   {\n" +
            "   }\n" +
            "}"
        ), new LineMarker.ToolTipEqualsAssert("Navigate to implementations"));
    }

    public void testThatLineMarkerIsProvidedForAttributeClass() {
        assertLineMarker(PhpPsiElementFactory.createPsiFileFromText(getProject(), "<?php\n" +
            "namespace Doctrine\\ORM\\Mapping;\n" +
            "" +
            "{\n" +
            "   #[\\Attribute]\n" +
            "   class Embedded\n" +
            "   {\n" +
            "   }\n" +
            "}"
        ), new LineMarker.ToolTipEqualsAssert("Navigate to implementations"));
    }

    public void testThatNonAnnotationClassMustNotProvideLineMarker() {
        assertLineMarkerIsEmpty(PhpPsiElementFactory.createPsiFileFromText(getProject(), "<?php\n" +
            "namespace Doctrine\\ORM\\Mapping;\n" +
            "" +
            "{\n" +
            "   /**\n" +
            "   */\n" +
            "   class Embedded\n" +
            "   {\n" +
            "   }\n" +
            "}"
        ));
    }
}

File: src/test/kotlin/de/espend/idea/php/annotation/tests/navigation/AnnotationUsageLineMarkerProviderTest.kt
package de.espend.idea.php.annotation.tests.navigation

import com.jetbrains.php.lang.psi.PhpPsiElementFactory
import de.espend.idea.php.annotation.tests.AnnotationLightCodeInsightFixtureTestCase
import de.espend.idea.php.annotation.tests.AnnotationLightCodeInsightFixtureTestCase.LineMarker.ToolTipEqualsAssert

/**
 * @author Daniel Espendiller <daniel@espendiller.net>
 */
class AnnotationUsageLineMarkerProviderTest : AnnotationLightCodeInsightFixtureTestCase() {
    @Throws(Exception::class)
    override fun setUp() {
        super.setUp()
        myFixture.copyFileToProject("AnnotationUsageLineMarkerProvider.php")
    }

    override fun getTestDataPath(): String =
        "src/test/kotlin/de/espend/idea/php/annotation/tests/navigation/fixtures"

    fun testThatLineMarkerIsProvidedForAnnotationClass() {
        assertLineMarker(
            PhpPsiElementFactory.createPsiFileFromText(
                getProject(), "<?php\n" +
                        "namespace Doctrine\\ORM\\Mapping;\n" +
                        "" +
                        "{\n" +
                        "   /**\n" +
                        "   * @Annotation\n" +
                        "   */\n" +
                        "   class Embedded\n" +
                        "   {\n" +
                        "   }\n" +
                        "}"
            ), ToolTipEqualsAssert("Navigate to implementations")
        )
    }

    fun testThatLineMarkerIsProvidedForAttributeClass() {
        assertLineMarker(
            PhpPsiElementFactory.createPsiFileFromText(
                getProject(), "<?php\n" +
                        "namespace Doctrine\\ORM\\Mapping;\n" +
                        "" +
                        "{\n" +
                        "   #[\\Attribute]\n" +
                        "   class Embedded\n" +
                        "   {\n" +
                        "   }\n" +
                        "}"
            ), ToolTipEqualsAssert("Navigate to implementations")
        )
    }

    fun testThatNonAnnotationClassMustNotProvideLineMarker() {
        assertLineMarkerIsEmpty(
            PhpPsiElementFactory.createPsiFileFromText(
                getProject(), "<?php\n" +
                        "namespace Doctrine\\ORM\\Mapping;\n" +
                        "" +
                        "{\n" +
                        "   /**\n" +
                        "   */\n" +
                        "   class Embedded\n" +
                        "   {\n" +
                        "   }\n" +
                        "}"
            )
        )
    }

    fun testThatNonAnnotationClassDoesNotAbortFollowingAnnotationClass() {
        assertLineMarker(
            PhpPsiElementFactory.createPsiFileFromText(
                getProject(), "<?php\n" +
                        "namespace Doctrine\\ORM\\Mapping;\n" +
                        "" +
                        "{\n" +
                        "   class PlainClass\n" +
                        "   {\n" +
                        "   }\n" +
                        "\n" +
                        "   /**\n" +
                        "   * @Annotation\n" +
                        "   */\n" +
                        "   class Embedded\n" +
                        "   {\n" +
                        "   }\n" +
                        "}"
            ), ToolTipEqualsAssert("Navigate to implementations")
        )
    }
}
